### ISSUE TITLE
Accelerate DCT and IDCT algorithms using AVX2 instructions

### ISSUE THREAD

**Issue by dcommander** [libjpeg-turbo/libjpeg-turbo#2]
Newer Intel/AMD chips provide support for AVX2 (256-bit integer SIMD) instructions, so developing new versions of the libjpeg-turbo SIMD algorithms that take advantage of this technology should improve the performance of libjpeg-turbo (although by how much is unknown.)  Because AVX2 doesn't increase the number of available registers, the best approach is probably to continue using hand-tuned NASM code in order to avoid unexpected register exhaustion.  256-bit SIMD would allow the row multiply operations in the DCT/IDCT algorithms to be executed in a single instruction instead of 3 or 4, and it will allow twice the data parallelism in the color conversion and up/downsampling routines.  What effect this will have on overall performance is unknown.  Huffman coding takes up something like 40-50%, so only 50-60% is available for optimization.  My gut says that we can probably achieve an overall performance boost of about 30% relative to the current SSE2 implementation, which would mean that libjpeg-turbo would be 2.5-7x as fast as libjpeg rather than 2-5x as fast.

**Comment by dcommander** [comment]
Thanks to a generation donation from [Userful](http://userful.com/), I now have a machine that supports AVX2 instructions.  I spent a couple of days in June doing the initial work of adding AVX2 support to the libjpeg-turbo code, and in the process of doing this, I created an AVX2 version of the simplest algorithm (integer quantization.)  I verified that the new AVX2-encoded version of that function is actually being invoked and is producing mathematically correct results, but it's actually slower than its SSE2 counterpart, despite having half the number of instructions.  Further investigation reveals that modern CPU architectures can only read/write 128 bits max at a time, so populating a 256-bit AVX2 register takes more than double the amount of time as it takes to populate a 128-bit SSE2 register. Unless there's enough intermediate compute going on to make up for it, then the overall performance will be slower, despite the compute portion being faster.  The next issue that arises is the fact that the DCT/IDCT algorithms and associated sample conversion algorithm operate on DCT rows and need to compute each row separately.  This is perfect for SSE2, because DCT coefficients are 16-bit when compressing an 8-bit JPEG, and each row is 8 coefficients wide (so 16 bits x 8 = 128 bits.)  Not so perfect for AVX2.  Even discounting the overhead of populating the 256-bit registers, I'm not sure what to do with them once they're populated.  The DCT algorithms tend to be column-parallel, meaning that you can perform the same operation on all 8 coefficients in each row, but the algorithms are not row-parallel.  Different rows are added to each other or subtracted from each other to compute intermediate values, some of which are multiplied by constants and others of which aren't, etc.  Not really any easy way to extend that to 256-bit.

This is not to say that AVX2 is useless.  It might be possible to, for instance, accelerate the matrix transpose operation that occurs frequently in the DCT kernels.  It is likely possible to accelerate the color conversion routines as well, since they are more embarrassingly parallel.  AVX2 would certainly provide performance benefits for 12-bit JPEG compression as well, since those algorithms use 32-bit DCT coefficients.  However, I think the takeaway is that this feature will not be as much of a "slam dunk" as previously thought.

One possible line of research would be to examine some of the other codec projects that are using AVX2 and see how they are leveraging the technology.

**Comment by dcommander** [comment]
Pushed to libjpeg-turbo 1.6

**Comment by mayeut** [comment]
Here's the "trivial" implementation for `jsimd_quantize_avx2` I'm talking about in 36df9d8e556b2347f3174bc772751747097934b9. Notice the `vzeroupper` instruction before `uncollect_args`

```
EXTN(jsimd_quantize_avx2):
    push        rbp
    mov         rax, rsp
    mov         rbp, rsp
    collect_args 3

    vmovdqu     ymm4, [XMMBLOCK(0,0,r12,SIZEOF_DCTELEM)]
    vmovdqu     ymm5, [XMMBLOCK(2,0,r12,SIZEOF_DCTELEM)]
    vmovdqu     ymm6, [XMMBLOCK(4,0,r12,SIZEOF_DCTELEM)]
    vmovdqu     ymm7, [XMMBLOCK(6,0,r12,SIZEOF_DCTELEM)]
    vpabsw      ymm0, ymm4
    vpabsw      ymm1, ymm5
    vpabsw      ymm2, ymm6
    vpabsw      ymm3, ymm7

    vpaddw      ymm0, ymm0, YMMWORD [CORRECTION(0,0,r11)]  ; correction + roundfactor
    vpaddw      ymm1, ymm1, YMMWORD [CORRECTION(2,0,r11)]
    vpaddw      ymm2, ymm2, YMMWORD [CORRECTION(4,0,r11)]
    vpaddw      ymm3, ymm3, YMMWORD [CORRECTION(6,0,r11)]
    vpmulhuw    ymm0, ymm0, YMMWORD [RECIPROCAL(0,0,r11)]  ; reciprocal
    vpmulhuw    ymm1, ymm1, YMMWORD [RECIPROCAL(2,0,r11)]
    vpmulhuw    ymm2, ymm2, YMMWORD [RECIPROCAL(4,0,r11)]
    vpmulhuw    ymm3, ymm3, YMMWORD [RECIPROCAL(6,0,r11)]
    vpmulhuw    ymm0, ymm0, YMMWORD [SCALE(0,0,r11)]       ; scale
    vpmulhuw    ymm1, ymm1, YMMWORD [SCALE(2,0,r11)]
    vpmulhuw    ymm2, ymm2, YMMWORD [SCALE(4,0,r11)]
    vpmulhuw    ymm3, ymm3, YMMWORD [SCALE(6,0,r11)]

    vpsignw     ymm0, ymm0, ymm4
    vpsignw     ymm1, ymm1, ymm5
    vpsignw     ymm2, ymm2, ymm6
    vpsignw     ymm3, ymm3, ymm7

    vmovdqu     [XMMBLOCK(0,0,r10,SIZEOF_DCTELEM)], ymm0
    vmovdqu     [XMMBLOCK(2,0,r10,SIZEOF_DCTELEM)], ymm1
    vmovdqu     [XMMBLOCK(4,0,r10,SIZEOF_DCTELEM)], ymm2
    vmovdqu     [XMMBLOCK(6,0,r10,SIZEOF_DCTELEM)], ymm3

    vzeroupper

    uncollect_args 3
    pop         rbp
    ret
```

With this function & `vzeroupper` instruction added in huffman encoding, here's what I get:
```
MacBook-Pro-de-Matthieu:build Matt$ JSIMD_FORCESSE2=1  ./tjbench ../nightshot_iso_100.ppm 95 -nowrite -subsamp 444 -componly

>>>>>  BGR (Top-down) <--> JPEG 4:4:4 Q95  <<<<<

Image size: 3136 x 2352
Compress      --> Frame rate:         15.912690 fps
                  Output image size:  2040975 bytes
                  Compression ratio:  10.841689:1
                  Throughput:         117.369964 Megapixels/sec
                  Output bit stream:  259.819218 Megabits/sec

MacBook-Pro-de-Matthieu:build Matt$ ./tjbench ../nightshot_iso_100.ppm 95 -nowrite -subsamp 444 -componly

>>>>>  BGR (Top-down) <--> JPEG 4:4:4 Q95  <<<<<

Image size: 3136 x 2352
Compress      --> Frame rate:         17.802981 fps
                  Output image size:  2040975 bytes
                  Compression ratio:  10.841689:1
                  Throughput:         131.312509 Megapixels/sec
                  Output bit stream:  290.683514 Megabits/sec
```

With only the `vzeroupper`instructions added:
```
MacBook-Pro-de-Matthieu:build Matt$ ./tjbench ../nightshot_iso_100.ppm 95 -nowrite -subsamp 444 -componly

>>>>>  BGR (Top-down) <--> JPEG 4:4:4 Q95  <<<<<

Image size: 3136 x 2352
Compress      --> Frame rate:         17.582069 fps
                  Output image size:  2040975 bytes
                  Compression ratio:  10.841689:1
                  Throughput:         129.683089 Megapixels/sec
                  Output bit stream:  287.076502 Megabits/sec
```

With no modifications:
```
MacBook-Pro-de-Matthieu:build Matt$ ./tjbench ../nightshot_iso_100.ppm 95 -nowrite -subsamp 444 -componly

>>>>>  BGR (Top-down) <--> JPEG 4:4:4 Q95  <<<<<

Image size: 3136 x 2352
Compress      --> Frame rate:         10.079649 fps
                  Output image size:  2040975 bytes
                  Compression ratio:  10.841689:1
                  Throughput:         74.346201 Megapixels/sec
                  Output bit stream:  164.578492 Megabits/sec
```

**Comment by dcommander** [comment]
Your modifications to the quantizer + the vzeroupper instruction were good for a few more percent compression performance across the board.  Those have been pushed to dev.  I was able to use the Intel SDE tool to verify that there are no remaining AVX-SSE transitions.

AVX2 Huffman encoding is still a wash compared to SSE2.  I pushed a modified version of that to the avx2_failed_experiments branch if you want to look at it further, but I don't consider it a priority.

**Comment by dcommander** [comment]
NOTE: Current speedup of the dev branch relative to master (1.5.x) on my Mac Mini/3.0 GHz i7:

**Comment by mayeut** [comment]
I checked the Huffman encoding but I couldn't do better than the SSE2 version. 

The CPU detection code is missing the check for operating system support as can be seen in https://software.intel.com/en-us/blogs/2011/04/14/is-avx-enabled

If I'm not mistaken, CentOS 5.x does not support AVX extensions (context switch do not save/restore YMM registers). Windows previous to Windows 7 SP1 does not support it either.

**Comment by denji** [comment]
Translate adaptive optimizer? https://github.com/kobalicek/asmjit#features & https://github.com/herumi/xbyak

**Comment by dcommander** [comment]
@denji no idea why those are relevant to this discussion.  Please clarify.

**Comment by dcommander** [comment]
@mayeut that should now be fixed in dev.

**Comment by mayeut** [comment]
@dcommander, the `xgetbv` check is still missing.

**Comment by dcommander** [comment]
I couldn't find any firm documentation on xgetbv, but as far as I can determine, it isn't necessary to do that check, because we aren't using floating point AVX (?)

**Comment by mayeut** [comment]
The {{xgetbv}} is necessary.

The OSXSAVE cpuid checks wether the OS uses "extended state management" with XSAVE/XRSTOR
The {{xgetbv}} check is here to check which states are managed through XSAVE/XRSTOR, as configured by the OS (x87/XMM/YMM).

Here's another PDF link from Intel. Chapter 2.2 & 3.2 are of interest for this specific issue.
https://software.intel.com/sites/default/files/4f/5b/36945

**Comment by dcommander** [comment]
Added the xgetbv check and pushed to dev

**Comment by dcommander** [comment]
This has been partially implemented (everything but the DCT/IDCT functions) in dev and currently provides the speedups listed above.

Adding the remaining algorithms would provide the following speedups relative to libjpeg-turbo 1.5.x:

18-32% (avg. 25%) speedup for 64-bit compression
15-29% (avg. 23%) speedup for 64-bit decompression

Since the initial funding for the project ran out, I have marked this as "funding needed" in hopes that someone else will step up to fund:
* Integration and cleanup of the DCT/IDCT algorithms.  Since these algorithms require multi-block versions of the underlying SIMD functions, I will need to investigate how best to interface those functions with the libjpeg C code-- preferably in a way that easily allows for future expansion (AVX512) and reuses as much of the existing code as possible.  The submitted implementations work but are not as elegant as I would prefer for a library that has this much exposure.
* Porting the algorithms to 32-bit
* Benchmarking and extensive regression testing

**Comment by nikolain** [comment]
I would like to help progress this issue a bit further to reach completion. Any details on types of funding / hardware / other help that might be useful to get DCT/IDCT AVX2 complete?

**Comment by dcommander** [comment]
I have the necessary hardware, but based on my experience thus far with integrating the color conversion and sampling routines, the labor to integrate the DCT/IDCT algorithms is likely to be extensive.  I'd like to secure US$3000-3500 to cover it.

**Comment by nikolain** [comment]
I think I can find about $1500.00 to help this started (and since I like libjpeg-turbo so much).  Hopefully, somebody else might help with second part of funding. I only care about x64 version of it to start with, if it helps in any way.

I'll contact you over professional services contacts for offline details once I know for sure.

**Comment by dcommander** [comment]
Any updates?

**Comment by des256** [comment]
Hi, I did a 16-bit AVX2 quantize+IDCT for a side project. I also have YUV2RGB code, but that is rather straightforward if you have MMX and SSE2. I'd like to donate it to libjpeg-turbo, but there might be conversion and compatibility issues. If anyone is interested, email me at desmond@germansmedia.nl.

**Comment by dcommander** [comment]
@des256 AVX2 implementations of YUV-to-RGB conversion and integer quantization have already been checked into the dev branch.  The IDCT algorithm has been implemented as well, by Intel, but it's going to take some effort to merge it.  Fortunately, I finally secured funding for that effort.  Since the Intel code is already written to be compatible with libjpeg-turbo, it makes the most sense to work with that code, at least for now.

That being said, if you have time, please look at our existing implementations (the IDCT algorithm is part of #30.)  If your implementation is better in some ways, then I'd be happy to borrow from it.

**Comment by des256** [comment]
I checked the implementations. My IDCT is comparable to jidctfst-sse2.asm, but exploiting AVX2. For sampling/color conversion, I made specific routines for a limited set of input encodings (Y-only, YCC420, YCC422, YCC440, YCC444 and RGB444) to ARGB8 output. That seems to be too limited for libjpeg-turbo. OTOH, you could consider extra speedy code paths for specific encodings.

**Comment by dcommander** [comment]
The Intel IDCT code (https://github.com/huaqiangwang/libjpeg-turbo/blob/avx2-dev/simd/jidctint-avx2-64.asm, https://github.com/huaqiangwang/libjpeg-turbo/blob/avx2-dev/simd/jidctfst-avx2-64.asm) is just a direct port of jidctint-sse2.asm and jidctfst-sse2.asm with AVX2 instructions, so it sounds like your implementation is very similar.  That code from Intel should integrate smoothly.  It's mainly the forward DCTs that are going to be problematic, because they require adding multi-block support to the underlying libjpeg architecture.

As far as color conversion, yes, the libjpeg architecture necessitates separating YCC-to-RGB conversion and upsampling, unless you're using "merged" (non-fancy) upsampling, in which case the algorithms are combined.  However, merged upsampling isn't the default, since it produces more artifacts than fancy upsampling.  Also, because merged upsampling is already faster than fancy upsampling, there isn't a lot of demand for further accelerating it.  Let me finish integrating the Intel code, then we can benchmark it against your code and determine whether it makes sense to provide further accelerations based on your algorithms.

**Comment by dcommander** [comment]
@des256 I'd like to revisit using some of your code.  It turns out that my assertion was incorrect regarding the Intel AVX2 IDCT implementation.  As with their AVX2 FDCT implementation, it requires extensive and disruptive modifications to the libjpeg C code, which makes it difficult to integrate (refer to #30.)  Thus, I am interested in any approaches that accelerate the SIMD modules using AVX2 instructions without fundamentally changing the data flow of those SIMD modules.

**Comment by des256** [comment]
Hi,

IDCT and FDCT are essentially the same on a low level. One 'trick' I 
used is to optimize the flow is reorder the pixels beforehand to reduce 
the instructions in the actual DCT code. There is another reordering 
step elsewhere (zigzag/dezigzag) that can be combined with this. I guess 
that might mess up your data flow.

I'm currently busy with something else, but I could probably help out in 
a week or two.

Best,
Desmond.

DRC schreef op 2018-01-29 12:56:

**Comment by dcommander** [comment]
It doesn't mess up the data flow as long as the reordering is confined to the DCT or IDCT function.  Is your code available somewhere that I could review it?

**Comment by des256** [comment]
Attached is the DCT code, it's currently embedded in another project, 
but it shouldn't be too hard to properly connect to libjpeg-turbo.

Best,
Desmond.


DRC schreef op 2018-01-29 15:03:

**Comment by des256** [comment]
Also, this code in particular is for 16-bit integers. For 32-bit 
integers and floats it'll look more like the SSE code.


DRC schreef op 2018-01-29 15:03:

**Comment by dcommander** [comment]
The attachment didn't come through.  Can you e-mail it to me?
https://libjpeg-turbo.org/About/Contact

**Comment by dcommander** [comment]
Since Intel's multi-block DCT/IDCT approach is too disruptive and not in keeping with the design principles of libjpeg-turbo, and since the nature of the DCT algorithm (with 16-bit coefficients) doesn't naturally lend itself to instruction sets larger than 128-bit, I spent some time this week looking at other possible ways to optimize the forward DCT and sample conversion algorithms using AVX2 instructions.

Although AVX2 instructions allowed me to implement the convsamp algorithm without a loop, I wasn't able to extract any signficant speedup.  Unfortunately, this algorithm seems limited by I/O, since it has to load eight 8-byte rows from non-contiguous memory locations.  I tried using AVX2 instructions to implement the matrix transpose in the forward DCT algorithm, but that slowed things down.  I also tried porting the forward DCT code to use VEX.128 AVX2 instructions (which work like three-operand equivalents of SSE2 instructions.)  This allowed me to reduce the instruction count significantly and optimize register usage, but unfortunately it did not produce more than about 1-2% speedup.  The VEX.128 and convsamp experiments have been checked into the avx2_failed_experiments branch.

I'm stumped on this unless there are other ideas, or until I can look at Desmond's code.

**Comment by dcommander** [comment]
In the spirit of incrementalism, I figured out a way to squeeze additional performance from the slow integer forward DCT algorithm by using AVX2 instructions to combine operations and eliminate stack swaps.  With the addition of the AVX2 convsamp algorithm, this boosts the compression speedup for 64-bit code from 7.5-28% (avg. 15%) to 13-36% (avg. 22%) and for 32-bit code from 5.8-13% (avg. 10%) to 10-21% (avg. 16%.)  It should be straightforward to do likewise for the slow integer inverse DCT algorithm.  Still trying to figure out how to speed up the fast integer DCT/IDCT.

**Comment by dcommander** [comment]
I've given up on the IFAST algorithms.  I can't seem to find any sane way to accelerate those using AVX2 instructions, but both of my failed experiments are now in the "avx2_failed_experiments" branch.  The IFAST DCT/IDCT is something of a legacy feature anyhow.  ISLOW is now as fast as IFAST on AVX2-equipped hardware.

I'm proceeding with accelerating the ISLOW IDCT, using the same techniques I used on the ISLOW FDCT.  Once that is completed, I will consider this issue closed, although I encourage others to review my work and try to squeeze additional performance out of it.

**Comment by dcommander** [comment]
The AVX2-accelerated ISLOW IDCT has been pushed to the dev branch.  It boosts the speedup for 64-bit YCbCr-to-RGB decompression from 4-26% (avg. 11%) to 9-35% (avg. 17%) on my machine, and the speedup for 32-bit YCbCr-to-RGB decompression from 4-13% (avg. 8%) to 7-17% (avg. 12%.)  That's as good as I can do for now, but again, I encourage others to review my work and try to improve upon it.  Also, if anyone can figure out how to accelerate the "fast" DCT/IDCT algorithms using AVX2, that would be very helpful.  I'm out of funding and thus can't spend any more time on this.

**Comment by twest820** [comment]
Hi, sort of an offshoot of this issue. Skimming through the assembly in master I'm not spotting VEX instructions but I see there are some VEX.128 commits in the AVX2 experimental branch. Not sure github's repo search is entirely trustworthy here but am I correct in inferring 1.5.3 doesn't do CPU dispatch to take advantage of VEX.128 on VEX capable processors? If so, is there a chance 1.6 might do this?

In my own code I support such dispatch as it's an easy way to get a 10% or so speedup (I'm using intrinsics, so it's just copy/paste into a VEX enabled translation unit) but am noticing some interesting x64 scalability differences between workloads which stay entirely in my VEX.128 versus ones which call into libjpeg-turbo. I've some hypothesis as to why this might be but need a clearer understanding of libjpeg-turbo's build disposition tell if there's anything to them.

In general, yeah, actually getting speedup from AVX's two lanes is frequently difficult. Often the CPI is double VEX.128, meaning computations take more wall clock time since the processor stays out of turbo.

**Comment by dcommander** [comment]
Your inference is correct.  1.5.x and prior releases don't use any instruction sets newer than plain vanilla SSE2.  I tried using VEX.128 to speed up the fast integer forward DCT (https://github.com/libjpeg-turbo/libjpeg-turbo/blob/da476915a25803cf28e820b5cc5dc5bee9f6ae47/simd/x86_64/jfdctfst-avx2.asm), but it wasn't any faster than the SSE2 version, which is why that code is in the avx2_failed_experiments branch.  The only VEX-capable CPU I have is a 3 GHz Core i7 (Haswell), so I'm not sure whether my inability to get any speedup from VEX.128 is due to the CPU.

The AVX2 code is all in the dev branch at the moment.  The master branch always points to the latest stable code, so dev will be merged into master when 1.6 beta1 is ready to release (and what's currently in master will be moved to a new branch called "1.5.x".)

The only remaining SSE2 SIMD algorithms that haven't been ported to AVX2 are:
* Fast integer forward DCT (see above regarding failed experiment)
* Fast integer inverse DCT
* 2x2 (1/4 Scaling) Integer Inverse DCT (infrequently used, and I don't care enough about it to spend any time porting it)
* 4x4 (1/2 Scaling) Integer Inverse DCT (infrequently used, and I don't care enough about it to spend any time porting it)
* Huffman encoding (neither Matthieu-- the original author of the SSE2 Huffman encoder-- nor I were able to accelerate it using AVX2, but this is also in the avx2_failed_experiments branch.  We didn't try to convert it to VEX.128.)

That's interesting about the dual-lane operations requiring double the CPI, and it makes sense based on my experience.  With the ISLOW FDCT/IDCT algorithms, I had to really restructure the algorithms to significantly reduce the instruction count and optimize the register usage (i.e. eliminate the use of the stack) before I could extract any speedup relative to SSE2.  Those optimizations wouldn't have been possible without using both lanes, however.  I also discovered that the vperm2i128 instruction, which is often necessary in order to make the algorithms work across two lanes, really slows things down.  I had to completely revisit parts of the algorithms in order to make them more dual-lane-friendly and thus reduce the use of vperm2i128 as much as possible.

**Comment by twest820** [comment]
Cool, thanks. That means I've some evidence suggesting VEX.128 would allow libjpeg-turbo to use hardware more efficiently than either SSE2 or AVX2 in certain cases when processing two or more jpegs concurrently. This might be an incorrect interpretation of the data, however, so let me check with Intel and evaluate some additional test cases. If correct it's potentially a processor specific behaviour and likely not so useful. I'll follow up in a few days.

In my evaluations of AVX2 vperm2i128 has had a way of being performance doom. Used sparingly in algorithms with heavy computation kernels it may be acceptable but lately it's been lighter codes I've been working with. In those a single lane swap could do things like make test cases which were otherwise 40% faster than VEX.128 run 15% slower. In evaluation, I've often had to do CPU dispatching on memory layouts to get advantage from 256 bit instructions. Perf has to be pretty important to justify maintaining two different sets of classes just to do the same thing slightly faster.

Something also to be aware of is while Haswell starts 256 bit wide execution immediately subsequent architectures take some hundreds of microseconds before the upper lane comes online. During this time both lanes are multiplexed to the low lane and throughput can't be any greater than VEX.128 (analogous to AVX on Sandy or Ivy Bridge). Depending on the algorithm it might be slower.

**Comment by dcommander** [comment]
As far as Skylake, it would be simple to insert a dummy instruction, as they suggest, to keep the upper lanes alive, but obviously that sort of optimization is impossible without testing the code on the actual chip.

**Comment by twest820** [comment]
Yeah, when I started developing on AVX and AVX2 I was expecting a general solution like SSE2 to SSE 4.2. That's not what it iss. It's a compute hammer for algorithms with loose coupling between points in memory and significant locality. In the image and other data processing tasks I've worked on what VTune shows is SSE/VEX.128 suffice to max out the DRAM channels when running parallel. So pretty much all VEX.256 accomplishes is to hammer the memory bottleneck harder. The increase in 8th gen core counts (Kaby Lake refresh, Coffee Lake) intensifies this problem as L3 size and DRAM bandwidth don't increase commensurately. So I tend to see AVX2 as something which lets you get really close to the hardware. A corollary of that is you're coding right up the physical limits of what's possible with the platform. Return on effort reduces accordingly.

I've often found SSSE3 and SSE4.1 instructions valuable agumentation to SSE2, whether VEX encoded or not. If AVX2 is picking up meaningful benefit from 256 bit versions of these it might be interesting to consider something of a backport. My availability's likely also limited, though, and some months out. We'll see where things land.

Wish there was a way to use the upper lane of ymm registers as xmm16-31. It'd save occasional contortions to avoid register spilling. This can sort of be achieved with data interleaves in memory which match the two lane architecture but in many cases I look at the cost of interleaving and deinterleaving exceeds the AVX2 compute advantage. Could get around this were true gather/scatter implemented but those instructions aren't there yet.

**Comment by dcommander** [comment]
I did a quick & dirty experiment, replacing my AVX2 version of the ISLOW FDCT with a VEX.128 version (basically a dumb port of the SSE2 code, but taking advantage of the three-operand instructions to eliminate all of the inter-register moves.)  The performance was identical to the SSE2 version, which is consistent with my previous findings with the IFAST FDCT.  Bearing in mind that my AVX2 version of the ISLOW FDCT involved some fundamental changes to the algorithm to make it more dual-lane friendly, I tried doing the same thing with the color conversion algorithm (dumb port from SSE2 to VEX.128, taking advantage of three-operand instructions.)  This algorithm was contributed by Intel and has a lot of vperm2i128 instructions.  Same results, however:  VEX.128 performed identically to SSE2, and the dual-lane AVX2 version was somewhat faster.

At least on Haswell, it doesn't appear possible to improve upon SSE2 by using VEX.128.  More generally, I'm not interested in bothering with a VEX.128 SIMD implementation unless it can be shown to be 10% or more faster than both AVX2 and SSE2 on a particular platform.  Otherwise, we need to just drop back to SSE2 on any CPU that has a slow dual-lane implementation.  Figuring out which CPUs that applies to is left as an exercise for the reader.
